xnmt/translator.py

```diff
@@ -191,46 +244,29 @@ def html_report(self, context=None):
       att_text.text = "Attention:"
       etree.SubElement(attention, 'br')
       attention_file = u"{}.attention.png".format(path_to_report)
-
-      if type(att) == dy.Expression:
-        attentions = att.npvalue()
-      elif type(att) == list:
-        attentions = np.concatenate([x.npvalue() for x in att], axis=1)
-      elif type(att) != np.ndarray:
-        raise RuntimeError("Illegal type for attentions in translator report: {}".format(type(attentions)))
-      xnmt.plot.plot_attention(src, trg, attentions, file_name = attention_file)
+      xnmt.plot.plot_attention(src, trg, att, file_name = attention_file)
 
     # return the parent context to be used as child context
     return html
 
   @handle_xnmt_event
   def on_file_report(self):
-    idx, src, trg, att = self.get_report_input()
-    path_to_report = self.get_report_path()
-
-    # Writing attention matrix to text file
-    if not any([src is None, trg is None, att is None]):
-      attention_file = u"{}.attention.txt".format(path_to_report)
-
-      if type(att) == dy.Expression:
-        attentions = att.npvalue()
-      elif type(att) == list:
-        attentions = np.concatenate([x.npvalue() for x in att], axis=1)
-      elif type(att) != np.ndarray:
-        raise RuntimeError("Illegal type for attentions in translator report: {}".format(type(attentions)))
-      assert attentions.shape == (len(src), len(trg))
-
-      # Not sure why elsewhere in the code the attention matrix is line-indexed by src
-      # and column-indexed by trg (the convention seems to be the other way around,
-      # eg. in [Bahdanau, 2014] https://arxiv.org/pdf/1409.0473.pdf)
-      # I expect attention matrices to sum to 1.0 line-wise.
-      att_transpose = np.transpose(attentions)
-      with io.open(attention_file, 'w', encoding='utf-8') as f:
-        f.write('\t' + '\t'.join(src) + '\n')
-        for i in range(len(trg)):
-          coeffs = '\t'.join(map(str, list(att_transpose[i])))
-          f.write('{}\t{}\n'.format(trg[i], coeffs))
-
+    idx, src, trg, attn = self.get_report_input()
+    assert attn.shape == (len(src), len(trg))
+    col_length = []
+    for word in trg:
+      col_length.append(max(len(word), 6))
+    col_length.append(max(len(x) for x in src))
+    with io.open(self.get_report_path() + ".attention.txt", encoding='utf-8', mode='w') as attn_file:
+      for i in range(len(src)+1):
+        if i == 0:
+          words = trg + [""]
+        else:
+          words = ["%.4f" % (f) for f in attn[i-1]] + [src[i-1]]
+        str_format = ""
+        for length in col_length:
+          str_format += "{:%ds}" % (length+2)
+        print(str_format.format(*words), file=attn_file)
```

Comment:
Hi Philip! Commenting here re: this change breaking things for me. If we merge it this way, I indeed will have to reintegrate somehow on a local branch the code I previously wrote, which is not a big deal. But just two thoughts:

* Shouldn't a report of the attention matrices be more agnostic of the particular formatting that either you or I prefer? Say using `numpy.savetxt` after adding a column and a line to have the source and target labels? If we just want to look at them it might be less readable, but to post-process them, it would be easier to do it this way in order to load the matrices in an external script. What do you think?
* My case for writing matrices where lines are indexed by the target and columns are indexed by the source, and that sum to 1.0 line-wise, is that it seems the standard convention, and certainly the convention in [(Bahdanau, 2014)](https://arxiv.org/pdf/1409.0473.pdf).

Reply:
Hi @pierregodard, thank you for the comment. I think the purpose of reporting is to provide users with an easy readable input, not to provide some formatting. The post-processable output should be a type of output, not a report. So adding one more field "attn_file" in addition to "hyp_file" in the translator is more appropriate I think. And you are right, it should be saved in binary format so that there is no loss of digit cancellation because floats cant be fully represented by string.

I think it is easier to sum it down, rather than to the right, but of course I can add option to make it line wise, probably tomorrow I will add this feature.